File: binary_tree/symmetric_tree.101.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// DFS solution
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        return isMirror(root->left, root->right);
        }
private:
    bool isMirror(TreeNode* left, TreeNode* right) {
        if (!left && !right) return true;
        if (!left || !right) return false;
        return (left->val == right->val) && isMirror(left->left, right->right) && isMirror(left->right, right->left);
    }
};

// BFS solution
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
    std::queue< std::pair<TreeNode*, TreeNode*> > nodeQueue;
    nodeQueue.push({root->left, root->right});
    while (!nodeQueue.empty()) {
        std::pair currentNode = nodeQueue.front();
        nodeQueue.pop();
        TreeNode* leftNode = currentNode.first;
        TreeNode* rightNode = currentNode.second;
        if (!leftNode && !rightNode ) continue;
        if (!leftNode || !rightNode ) return false;
        if (leftNode->val != rightNode->val) return false;
        nodeQueue.push({leftNode->left, rightNode->right} );
        nodeQueue.push({leftNode->right, rightNode->left});
    }
    return true;
    }
};

File: binary_tree/symmetric_tree.101.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Recursively
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        # base
        if not root:
            return True
        # recursion
        def dfs(left_node, right_node):
            if not left_node and not right_node:
                return True
            if not left_node or not right_node:
                return False
            if left_node.val != right_node.val:
                return False
            return dfs(left_node.left, right_node.right) and dfs(left_node.right, right_node.left)
        return dfs(root.left, root.right)
    
# Iteratively
